tests/test_training.py

```diff
@@ -114,6 +115,64 @@ def test_training(self):
         loss = np.mean(epoch_losses)
         self.assertFalse(np.isnan(loss))
 
+    def test_patch_wise_training(self):
+        """
+        Test model training with patch-wise tasks.
+        """
+        tl = TaskLoader(context=self.da, target=self.da)
+        model = ConvNP(self.data_processor, tl, unet_channels=(5, 5, 5), verbose=False)
+
+        # generate training tasks
+        n_train_dates = 10
+        dates = [np.random.choice(self.da.time.values) for i in range(n_train_dates)]
+        train_tasks = tl.generate_tasks(
+            dates,
+            context_sampling="all",
+            target_sampling="all",
+            patch_strategy="random",
+            patch_size=(0.8, 0.8),
+        )
+
+        # TODO pytest can also be more succinct with pytest.fixtures
+        # Train
+        trainer = Trainer(model, lr=5e-5)
+        batch_size = None
+        # TODO check with batch_size > 1
+        # batch_size = 5
+        n_epochs = 5
+        epoch_losses = []
+        for epoch in tqdm(range(n_epochs)):
+            batch_losses = trainer(train_tasks, batch_size=batch_size)
+            epoch_losses.append(np.mean(batch_losses))
+
+        # Check for NaNs in the loss
+        loss = np.mean(epoch_losses)
+        self.assertFalse(np.isnan(loss))
+
+    def test_sliding_window_training(self):
```

Comment:
I think the trainer tests could also benefit from fixtures, because at the moment there is a lot of code duplication. Maybe fixtures that generate a list of `train_tasks` that are then run with a `ConvCNP` and `Trainer`